### ISSUE TITLE
Videos flashing/cannot be played.

### ISSUE THREAD

**Issue by ghost** [luakit/luakit#990]
**Current Behavior:**
When playing Youtube (and Reddit) videos, the screen constantly flashes.
It seems that only VP9 ones have this problem; AV1 and H264 ones play fine.
Also, this problem only occurs on Youtube and Reddit. The videos on [VP9 test page](https://base-n.de/webm/VP9%20Sample.html) plays fine.
Sometimes the ad before the video can play, but after the ad, the video cannot be played.
I also turned off all the hardware rendering options and the problem persists.
Firefox can play those videos normally. There is no error or warning in the log when this happens.

https://user-images.githubusercontent.com/103848171/163686869-1795c0cc-b235-42a8-a7ed-4dc1fbaef850.mp4

**Desired Behavior:**
The video plays normally.

**How can we reproduce it (step by step):**
1. Install LuaKit (and other GStreamer libraries).
2. Play any YouTube/Reddit video.

**Environment:**

Linux Distribution & Version: Debian Sid
Output of `luakit --version`:

  luakit 2.3
    built with webkit 2.34.2 (installed version: 2.36.0)

**Note about webkit issues:**

I cannot test it since Ephiphany cannot use a proxy (the variables are ignored).

If you're reporting a rendering issue, please test it with the gnome
browser ephiphany as well. If the issue occurs there too, we're very
likely not able to help. These issues should be reported to webkit:
https://bugs.webkit.org

**Comment by taobert** [comment]
The flashing that i see is also present in Epiphany and Qutebrowser,
and is therefore likely a webkit(/whatever webkit includes) issue (and not something Luakit can/should correct/work around).
Can you confirm that other webkit browsers display the same behaviour?

**Comment by taobert** [comment]
Um.
Further, i see the same flashing if i download the video and play it with mpv.
vlc by contrast doesn't play it, but print's a bunch of errors:
```
00007f3688097cf0] main video output error: video output creation failed
[00007f36b8c1ccc0] main decoder error: failed to create video output
[h264 @ 0x7f36b8cd1000] get_buffer() failed
[h264 @ 0x7f36b8cd1000] thread_get_buffer() failed
[h264 @ 0x7f36b8cd1000] decode_slice_header error
[h264 @ 0x7f36b8cd1000] no frame!
```
...so my suspicion is that the video you posted is b0rked.
Can you provide some other examples?

**Comment by ghost** [comment]
Here is another example.

https://user-images.githubusercontent.com/103848171/163724964-68c855fe-1043-412b-bb93-9da087b35665.mp4

I cannot test other browsers (Epiphany refuses to use proxy, QuteBrowser does not play videos at all). There is [a bug](https://bugs.webkit.org/show_bug.cgi?id=176651) on WebKit's tracker but I doubt it is related.
I also downloaded some YT videos in VP9 format and played them using `gst-play-1.0` (since GStreamer is used for playing); all played without problem.

**Comment by taobert** [comment]
Ok. I had misunderstood your initial video as a problem video, rather than a video of the problem,
so my seeing the same flashing in other browsers/players is irrelevant.

When i look at [https://www.youtube.com/watch?v=DPJqCVj-Li0](https://www.youtube.com/watch?v=DPJqCVj-Li0) it plays fine for me.
Can you provide an actual youtube id/link which doesn't play for you?
Do the videos you downloaded play correctly if you open them directly in luakit (removing streaming/YT player effects)?

I'm also on Debian Sid, so that shouldn't be an issue, though we may have different packages installed.

**Comment by ghost** [comment]
> Can you provide an actual youtube id/link which doesn't play for you?

Every video using VP9 cannot be played.

> Do the videos you downloaded play correctly if you open them directly in luakit (removing streaming/YT player effects)?

Yes. Using `file://path/` plays the video correctly. (The videos on the VP9 test page plays as well.)
The problem seems to be a YouTube thing.

**Comment by c0dev0id** [comment]
Not sure if it helps, but can you try setting `export LIBGL_DRI3_DISABLE=1` before starting luakit? I'm not able to reproduce it, but I remember video issues on some intel chipsets a few years back.

**Comment by ghost** [comment]
I tried. Not working.
I am using an AMD Ryzen 5700G (iGPU).

**Comment by c0dev0id** [comment]
Hmm. It was worth a try. Sorry, then I can't really help :(
Btw. AMD Ryzen 7 PRO 4750U here.

**Comment by taobert** [comment]
Yes, i don't know either.
In lieu of being able to reproduce it locally, i'm down to spitballing.

You say Debian Sid, but confirm that the kernel and video drivers are up-to-date.
(I seem to recall a flashing thing some months ago, though it may have been nouveau specific, and i think it effected _all_ video.)

Do you have a user_agent set?
Since it seems to be the youtube player that's failing to work, it might be javascripty.
I don't know how uniform various browsers' implementations are,
but it's conceivable to me that the player might use different code for different browsers,
and might test the user_agent string to determine what to (attempt to) do.

You could trawl through the `apt show <package>` output(s),
starting with `luakit` and following the chain of anything listed as `Depends:`, `Recommends:` or `Suggested:`.
Depended-on packages are automatically installed, but `Recommended:` and `Suggested:` packages aren't.
Something that's only needed to make VP9 work without stuttering might well be in the latter group,
so make sure that any packages listed as `Recommended:` or `Suggested:` are also installed.
This is tedious.
I had a bit of a poke around down this road for you,
and didn't come up with anything that'd let me reproduce what you see, but i may have missed something.

You could try compiling yourself.
The Debian version's pretty up to date (and also works for me),
but `apt build-dep luakit` might drag in something that is not listed as a "Depends:",
but which is needed for the player to work well.
Likewise, the compilation might complain about something that gives you a hint.

You could try [Christian Marillat's](https://www.deb-multimedia.org) packages.
Largely for historical reasons, i have/had `deb http://www.deb-multimedia.org sid main non-free` in my `/etc/apt/sources.list`.
I note that this includes a number of gstreamer libs.
I think i've now removed all the deb-multimedia packages and reinstalled them with their main-repository versions,
and this hasn't allowed me to reproduce what you see,
but it's possible that some of the deb-multimedia packages will have slightly different behaviour and might give you a hint about what's going on.

Trying Epiphany from another location, which doesn't require a proxy,
would be useful confirmation that this is _not_ luakit specific,
though i assume that if this were trivial you would have already done it.

**Comment by ghost** [comment]
It seems that the problem magically fixed itself when I removed `gstreamer1.0-vaapi`. When reinstalling the problem reappeared.

Edit: [Found the bug.](https://bugs.webkit.org/show_bug.cgi?id=238728) I can close it now.
